lib/types/SSI.types.ts

```diff
@@ -55,21 +63,153 @@ export interface Issuer {
 }
 
 export interface Credential {
-  '@context': string[];
-  id: string;
-  type: string[];
+  getType(): string;
+
+  getBaseCredential(): CredentialBase;
```

Comment:
I see that the JSON-LD version returns this for getBaseCredential(). The name suggests that you can find everything here (base). But a JWT based credential really is the JWT payload standard claims:

- Encoded as standard JOSE header parameters, or
- Encoded as registered JWT claim names, or
- Contained in the JWS signature part.

together with the vc respective vp claims in the JWT. See https://www.w3.org/TR/vc-data-model/#json-web-token-extensions.


Then there are the following 'rules' 
> - alg MUST be set for digital signatures. If only the proof property is needed for the chosen signature method (that is, if there is no choice of algorithm within that method), the alg header MUST be set to none.
> - kid MAY be used if there are multiple keys associated with the issuer of the JWT. The key discovery is out of the scope of this specification. For example, the kid can refer to a key in a DID document, or can be the identifier of a key inside a JWKS.
> - typ, if present, MUST be set to JWT.
> - For backward compatibility with JWT processors, the following registered JWT claim names MUST be used instead of, or in addition to, their respective standard verifiable credential counterparts:
> - 
> - exp MUST represent the expirationDate property, encoded as a UNIX timestamp (NumericDate).
> - iss MUST represent the issuer property of a verifiable credential or the holder property of a verifiable presentation.
> - nbf MUST represent issuanceDate, encoded as a UNIX timestamp (NumericDate).
> - jti MUST represent the id property of the verifiable credential or verifiable presentation.
> - sub MUST represent the id property contained in the verifiable credential subject.
> - aud MUST represent (i.e., identify) the intended audience of the verifiable presentation (i.e., the verifier intended by the presenting holder to receive and verify the verifiable presentation).
> - Other JOSE header parameters and JWT claim names not specified herein can be used if their use is not explicitly discouraged. Additional verifiable credential claims MUST be added to the credentialSubject property of the JWT.

Doesn't it make sense that we have full interfaces for both a JWT Based Credential and a JSONLD based credential, but that we internally use a class that can accept either one of those and then make that class return the correct values from the respective types, a bit similar to what was done with the getContext. So we have an internal Representation of a (V)C and (V)P that is stable no matter the type.

So instead of having the inheritance and thus exposing our way of trying the make it uniform in the structure, remove the inheritance with the base credential exposure. Allow for both types to be passed in as arguments. But as a first step in our code wrap it in a class, that we know behaves in a uniform way. So for a JWT based VC it for instance it woud use exp for the expirationDate, whilst for a JSONLD version it would simply return the corresponding value.

I think that is cleaner for an end user and more future proof for us at the same time

Reply:
Ideally we would create 2 classes that implement a common interface for us internally. That one should be pretty close to the JSONLD version, but probably a bit more relaxed (JWT credentials don't necessarily have a '@context' for instance, as that is JSONLD). Maybe an 'internal' abstract class makes sense.

Then we have a JWTCredential class and a JSONLDCredential class implementing the internal interface or extending the abstract class.

Now we can call for instance `getSubjectId()`, which for JSONLD returns `credentialSubject.id`, and for JWT return the `sub`. Given we use the internal representation in our code everywhere, we do not access these properties directly but only using `getSubjectId()` etc.

Reply:
It would also allow for a nice serialization solution, since we can incorporate that into the respective classes.

Reply:
Also this link gives a lot of pointers on the actual mappings: https://w3c.github.io/vc-data-model/#jwt-decoding

Reply:
Lastly in the class we could have something like `originalPayload` which is the payload fitting the JSONLD and JWT interfaces the library users pass in, so that we can also alway access that from the class if there is a need (for instance for (de)serialization)

Reply:
I'm adding the methods to the Credential. But on a broader spec, I don't think it will effect our process at all. because whenever we're dealing with filters (on any thing it doesn't matter `issuance_date`, `issuer`, ..) we will use the `path` property that is defined in the filter.

Reply:
Yes, that is what I was referring to in the last answer. We work in a lot of cases directly on the json path. With having a stable entry to known properties I can forse that we deviate every now and then from that, like maybe for instance for the subject is issuer use cases etc.